server/controllers/finance/reports/budget_analytical/index.js

```diff
@@ -0,0 +1,235 @@
+const q = require('q');
+const _ = require('lodash');
+
+const ReportManager = require('../../../../lib/ReportManager');
+const db = require('../../../../lib/db');
+const Budget = require('../../budget');
+const Fiscal = require('../../fiscal');
+
+const BUDGET_REPORT_TEMPLATE = './server/controllers/finance/reports/budget_analytical/report.handlebars';
+
+// expose to the API
+exports.report = report;
+
+/**
+ * @function report
+ *
+ * @description
+ * Renders the analytical budget report.
+ */
+async function report(req, res, next) {
+  const params = req.query;
+  const fiscalYearId = parseInt(params.fiscal_id, 10);
+  const setNumberYear = parseInt(params.set_number_year, 10) + 1;
+  const colspanValue = parseInt(params.set_number_year, 10) + 6;
+  const hideUnused = parseInt(params.hide_unused, 10);
+  const { enterprise } = req.session;
+
+  let totalBudgetIncome = 0;
+  let totalRealisationIncome = 0;
+  let totalVariationIncome = 0;
+
+  let totalBudgetExpense = 0;
+  let totalRealisationExpense = 0;
+  let totalVariationExpense = 0;
+
+  const optionReport = _.extend(params, {
+    csvKey : 'rows',
+    renameKeys : false,
+    orientation : 'landscape',
+  });
+
+  try {
+    const transaction = [];
+    const reportColumn = [];
+    const reportFootColumIncome = [];
+    const reportFootColumExpense = [];
+
+    const fiscalYear = await Fiscal.lookupFiscalYear(fiscalYearId);
+
+    const sqlGetPreviousFiscalYear = `
+      SELECT fy.id, fy.label, YEAR(fy.end_date) AS year
+        FROM fiscal_year AS fy
+        WHERE fy.id <= ? ORDER BY fy.id DESC
+        LIMIT ?
+    `;
+
+    const fiscalsYear = await db.exec(sqlGetPreviousFiscalYear, [fiscalYearId, setNumberYear]);
+    const reporting = new ReportManager(BUDGET_REPORT_TEMPLATE, req.session, optionReport);
+
+    fiscalsYear.forEach((fisc, index) => {
+      transaction.push(Budget.buildBudgetData(fisc.id));
+
+      if (index > 0) {
+        // Getting the fiscal years for the report header columns
+        reportColumn.push({ number : fisc.year });
+      }
+    });
+
+    const fiscalYearNumber = fiscalsYear[0].year || '';
+    const lastFiscalYearNumber = fiscalsYear[1].year || '';
+
+    const dataFiscalsYear = await q.all(transaction);
+    let tabFiscalReport = [];
+    const uniqueSet = new Set();
+
+    dataFiscalsYear.forEach(subArray => {
+      subArray.forEach(item => {
+        const key = `${item.id}-${item.number}`;
+        if (!uniqueSet.has(key)) {
+          uniqueSet.add(key);
+          const isTitle = (item.type_id === 6);
+
+          if (item.type_id) {
+            tabFiscalReport.push(
+              {
+                id : item.id,
+                number : item.number,
+                label : item.label,
+                type_id : item.type_id,
+                isTitle,
+              });
+          }
+        }
+      });
+    });
+
+    dataFiscalsYear.forEach((fiscal, index) => {
```

Comment:
I am not sure why you are treating index === 0 as different.  I would advise you to just break it into two different functions/blocks though, for clarity:

```js
const [firstYear, ...laterYears] = dataFiscalsYear;

tabFiscalreport.forEach(rep => {
  if (firstYear.number === rep.number) {
     // do whatever 
   } 
}

// later:
tabFiscalreport.forEach(rep => {
  let totalIncomeRealised = 0;
  let totalExpenseRealised = 0;
   /// etc
 })
```
This saves you from a big `.forEach()` loop that just calls `if` first thing.

Reply:
I didn't quite understand what I should do, please give more guidance.

Reply:
> I am not sure why you are treating index === 0 as different. I would advise you to just break it into two different functions/blocks though, for clarity:

FYI If you check the SQL definition of the `budget` table, it indicates that period==0 is for the FY budget total for the account.

Reply:
After discussion with @jniles, I suggest we leave cleaning this up for a later PR (@jniles will create the PR).